src/Xamarin.Android.Build.Tasks/Utilities/MarshalMethodsClassifier.cs

```diff
@@ -0,0 +1,388 @@
+			public string TypeName                    { get; }
+			public AssemblyNameReference AssemblyName { get; }
+
+			public ConnectorInfo (string spec)
+			{
+				string[] connectorSpec = spec.Split (':');
+				MethodName = connectorSpec[0];
+
+				if (connectorSpec.Length < 2) {
+					return;
+				}
+
+				string fullTypeName = connectorSpec[1];
+				int comma = fullTypeName.IndexOf (',');
+				TypeName = fullTypeName.Substring (0, comma);
+				AssemblyName = AssemblyNameReference.Parse (fullTypeName.Substring (comma + 1).Trim ());
+			}
+		}
+
+		interface IMethodSignatureMatcher
+		{
+			bool Matches (MethodDefinition method);
+		}
+
+		sealed class NativeCallbackSignature : IMethodSignatureMatcher
+		{
+			static readonly HashSet<string> verbatimTypes = new HashSet<string> (StringComparer.Ordinal) {
+				"System.Boolean",
+				"System.Byte",
+				"System.Char",
+				"System.Double",
+				"System.Int16",
+				"System.Int32",
+				"System.Int64",
+				"System.IntPtr",
+				"System.SByte",
+				"System.Single",
+				"System.UInt16",
+				"System.UInt32",
+				"System.UInt64",
+				"System.Void",
+			};
+
+			readonly List<string> paramTypes;
+			readonly string returnType;
+			readonly TaskLoggingHelper log;
+
+			public NativeCallbackSignature (MethodDefinition target, TaskLoggingHelper log)
+			{
+				this.log = log;
+				returnType = MapType (target.ReturnType.FullName);
+				paramTypes = new List<string> {
+					"System.IntPtr", // jnienv
+					"System.IntPtr", // native__this
+				};
+
+				foreach (ParameterDefinition pd in target.Parameters) {
+					paramTypes.Add (MapType (pd.ParameterType.FullName));
+				}
+			}
+
+			string MapType (string type)
+			{
+				if (verbatimTypes.Contains (type)) {
+					return type;
+				}
+
+				return "System.IntPtr";
+			}
+
+			public bool Matches (MethodDefinition method)
+			{
+				if (method.Parameters.Count != paramTypes.Count || !method.IsStatic) {
+					log.LogDebugMessage ($"Method '{method.FullName}' doesn't match native callback signature (invalid parameter count or not static)");
+					return false;
+				}
+
+				if (String.Compare (returnType, method.ReturnType.FullName, StringComparison.Ordinal) != 0) {
+					log.LogDebugMessage ($"Method '{method.FullName}' doesn't match native callback signature (invalid return type)");
+					return false;
+				}
+
+				for (int i = 0; i < method.Parameters.Count; i++) {
+					ParameterDefinition pd = method.Parameters[i];
+					if (String.Compare (pd.ParameterType.FullName, paramTypes[i], StringComparison.Ordinal) != 0) {
+						log.LogDebugMessage ($"Method '{method.FullName}' doesn't match native callback signature, expected parameter type '{paramTypes[i]}' at position {i}, found '{pd.ParameterType.FullName}'");
+						return false;
+					}
+				}
+
+				return true;
+			}
+		}
+
+		TypeDefinitionCache tdCache;
+		DirectoryAssemblyResolver resolver;
+		Dictionary<string, IList<MarshalMethodEntry>> marshalMethods;
```

Comment:
Given that this is a private field, why `IList<MarshalMethodEntry>` and not just `List<MarshalMethodEntry>`?  Interface method invocation is (slightly) slower than class method invocation.

Reply:
The field is exported by the `MarshalMethods` property